src/OpenLoco/src/Windows/ObjectSelectionWindow.cpp

```diff
@@ -517,128 +505,116 @@ namespace OpenLoco::Ui::Windows::ObjectSelectionWindow
+
+            const bool subTabIsVisible = showSecondaryTabs && i < subTabs.size() && shouldShowSubTab(subTabs, i, FilterLevel(self.var_856));
+            if (!subTabIsVisible)
+                self.disabledWidgets |= 1ULL << widgetIndex;
+            else
+                self.disabledWidgets &= ~(1ULL << widgetIndex);
+
+            if (subTabIsVisible)
+                self.enabledWidgets |= 1ULL << widgetIndex;
+            else
+                self.enabledWidgets &= ~(1ULL << widgetIndex);
+
+            if (self.currentSecondaryTab == i)
+                self.activatedWidgets |= 1ULL << widgetIndex;
+            else
+                self.activatedWidgets &= ~(1ULL << widgetIndex);
         }
 
-        self.activatedWidgets &= ~((1ULL << widx::vehicleTypeTrain) | (1ULL << widx::vehicleTypeBus) | (1ULL << widx::vehicleTypeTruck) | (1ULL << widx::vehicleTypeTram) | (1ULL << widx::vehicleTypeAircraft) | (1ULL << widx::vehicleTypeShip));
+        Widget::leftAlignTabs(self, widx::secondaryTab1, widx::secondaryTab8, 30);
+
         if (showSecondaryTabs)
         {
-            self.activatedWidgets |= 1ULL << (widx::vehicleTypeTrain + self.currentSecondaryTab);
-            self.enabledWidgets |= (1ULL << widx::vehicleTypeTrain) | (1ULL << widx::vehicleTypeBus) | (1ULL << widx::vehicleTypeTruck) | (1ULL << widx::vehicleTypeTram) | (1ULL << widx::vehicleTypeAircraft) | (1ULL << widx::vehicleTypeShip);
-
-            self.widgets[widx::scrollview].top = 85 + 28;
+            self.widgets[widx::scrollview].top = 62 + 28;
             self.widgets[widx::scrollviewFrame].type = WidgetType::panel;
             self.widgets[widx::scrollviewFrame].top = self.widgets[widx::scrollview].top - 2;
-            self.widgets[widx::objectImage].top = 68 + 28;
         }
         else
         {
-            self.enabledWidgets &= ~((1ULL << widx::vehicleTypeTrain) | (1ULL << widx::vehicleTypeBus) | (1ULL << widx::vehicleTypeTruck) | (1ULL << widx::vehicleTypeTram) | (1ULL << widx::vehicleTypeAircraft) | (1ULL << widx::vehicleTypeShip));
-
-            self.widgets[widx::scrollview].top = 85;
+            self.widgets[widx::scrollview].top = 62;
             self.widgets[widx::scrollviewFrame].type = WidgetType::none;
-            self.widgets[widx::objectImage].top = 68;
         }
     }
 
     static loco_global<uint16_t[kMaxObjectTypes], 0x0112C1C5> _112C1C5;
     static loco_global<uint32_t, 0x0112C209> _112C209;
 
-    static constexpr uint8_t kRowOffsetX = 10;
-    static constexpr uint8_t kRowOffsetY = 24;
-
     // 0x0047328D
     static void drawTabs(Window* self, Gfx::DrawingContext& drawingCtx)
     {
-        auto y = self->widgets[widx::panel].top + self->y - 26;
-        auto x = self->x + 3;
+        auto yPos = self->y + self->widgets[widx::tabArea].top;
+        auto xPos = self->x + 3;
 
-        for (auto row = 1; row >= 0; row--)
+        for (auto index = 0; _tabPositions[index] != 0xFF; index++)
```

Comment:
if we use std::array, do you even need the final position to be 0xFF? since you can just iterate over array::length instead?

Reply:
Hmm, `std::array` is a fixed length, though. Come to think of it, I could've just made it a vector to make it dynamic. I agree that would be a good change. Let's go with that.

Reply:
ah yeah of course, vector is fine, I got confused when you mentioned array previously :|